---
title: "Class 18"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Hands-on Worksheet Part 1

#### The Genomic Data Commons R package

```{r}
library(GenomicDataCommons)
library(TCGAbiolinks)
library(maftools)
```

#### Querying the GDC from R

```{r}
cases_by_project <- cases() %>%
  facet("project.project_id") %>%
  aggregations()
head(cases_by_project)
```

```{r}
x <- cases_by_project$project.project_id
colvec <- rep("lightblue", nrow(x))
barplot(x$doc_count,names.arg=x$key,log="y",las=2,col=colvec)
```

#### Variant analysis with R

```{r}
maf.file <- GDCquery_Maf(tumor="PAAD", pipelines = "mutect")
vars = read.maf(maf = maf.file, verbose = FALSE)
lollipopPlot(vars, gene='KRAS')
```

## Hands-on Worsheet Part 2

#### Identifing sites of mutation

```{r}
library(bio3d)
```

```{r}
seqs <- read.fasta("seq.fa")
seqs
```

```{r}
## Calculate positional identity scores
ide <- conserv(seqs$ali, method="identity")
mutant.sites <- which(ide < 1) 

## Exclude gap possitions from analysis
gaps <- gap.inspect(seqs)
mutant.sites <- mutant.sites[mutant.sites %in% gaps$f.inds]
mutant.sites
```

```{r}
## Make a "names" label for our output sequences (one per mutant)
mutant.names <- paste0(seqs$ali["P53_wt",mutant.sites],mutant.sites,seqs$ali["P53_mutant",mutant.sites])
mutant.names
```

```{r}
## Sequence positions surounding each mutant site
start.position <- mutant.sites - 8
end.position <-  mutant.sites + 8

# Blank matrix to store sub-sequences
store.seqs <- matrix("-", nrow=length(mutant.sites), ncol=17)
rownames(store.seqs) <- mutant.names

## Extract each sub-sequence
for(i in 1:length(mutant.sites)) {store.seqs[i,] <- seqs$ali["P53_mutant",start.position[i]:end.position[i]]}
store.seqs
```

```{r}
## Output a FASTA file for further analysis
write.fasta(seqs=store.seqs, ids=mutant.names, file="subsequences.fa")
```